### ISSUE TITLE
ActiveRecord::QueryMethods includes is not working correctly with join

### ISSUE THREAD

**Issue by sromano** [rails/rails#16436]
# Using Rails 4.1.4

The queries are correct but the preloaded objects are not being actually associated

If I run

``` ruby
Style.j.includes(:product).all.each do |style|
  style.product.id
end
```

Then I get the following queries

``` sql
SELECT `styles`.* FROM `styles`
SELECT `products`.* FROM `products`  WHERE `products`.`id` IN (1, 2)
```

However, if I run, 

``` ruby
Style.joins('INNER JOIN display_category_positions ON display_category_positions.product_id = styles.product_id').includes(:product).all.each do |style|
  style.product.id
end
```

Then I get the following queries

``` sql
SELECT `styles`.* FROM `styles` INNER JOIN display_category_positions ON display_category_positions.product_id = styles.product_id
SELECT `products`.* FROM `products`  WHERE `products`.`id` IN (1, 2)
SELECT  `products`.* FROM `products`  WHERE `products`.`id` = 1 LIMIT 1  [["id", 1]]
SELECT  `products`.* FROM `products`  WHERE `products`.`id` = 2 LIMIT 1  [["id", 2]]
```

You can see that it is not preloading correctly the objects although it is querying them
If you want to know the code of the classes

``` ruby
class Style < ActiveRecord::Base
  belongs_to :product
  has_many :display_category_positions
end
```

``` ruby
class Product < ActiveRecord::Base
    has_many :styles
end
```

**Comment by messanjah** [comment]
I believe this illustrates the same issue in 5.1.3.

```ruby
# frozen_string_literal: true

begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"
  # Activate the gem you are reporting the issue against.
  gem "activerecord", "5.1.3"
  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# Ensure backward compatibility with Minitest 4
Minitest::Test = MiniTest::Unit::TestCase unless defined?(Minitest::Test)

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :authors, force: :true do |t|
  end

  create_table :posts, force: true do |t|
    t.integer :author_id
  end

  create_table :comments, force: true do |t|
    t.integer :post_id
  end
end

class Author < ActiveRecord::Base
  has_many :posts
end

class Post < ActiveRecord::Base
  has_many :comments
  belongs_to :author
end

class Comment < ActiveRecord::Base
  belongs_to :post
end

class BugTest < Minitest::Test
  def test_preload_association_cache_combined_with_joins
    first_author = Author.create!
    first_post = first_author.posts.create!
    first_post.comments.create!
    first_post.comments.create!

    second_author = Author.create!
    second_post = second_author.posts.create!
    second_post.comments.create!

    posts = Post.joins(:comments).preload(:author)

    posts.each do |post|
      assert post.association_cached?(:author)
    end
  end
end
```

**Comment by kamipo** [comment]
Looks like that this is a limitation for JOIN with has_many association.
`Post.joins(:comments).count` returns 3 (contained duplicated `first_post` by `joins(:comments)`) in that case.
`distinct` is required if we don't expect duplicated records when using `joins` with has_many association.

**Comment by kamipo** [comment]
Fixed by #33895.

**Comment by iBublik** [comment]
In our case issue was in model with non unique ID (under the hood it takes data from database view that unions multiple tables). Rails [uses ID here](https://github.com/rails/rails/blob/v5.0.7/activerecord/lib/active_record/core.rb#L435), which is used as object identity when object serves as key in a hash. So when query contained multiple records with same ID, associations was not linked as expected, because of using of object as hash key [here](https://github.com/rails/rails/blob/v5.0.7/activerecord/lib/active_record/associations/preloader/association.rb#L72) - associated object was linked only to last object within of group with duplicate IDs
